scapy/packet.py

```diff
@@ -138,8 +138,20 @@ def __init__(self, _pkt=b"", post_transform=None, _internal=0, _underlayer=None,
             self.dissect(_pkt)
             if not _internal:
                 self.dissection_done(self)
-        for f, v in six.iteritems(fields):
-            self.fields[f] = self.get_field(f).any2i(self, v)
+        # We use this strange initialization so that the fields
+        # are initialized in their declaration order.
+        # It is required to always support MultipleTypeField
+        for field in self.fields_desc:
```

Comment:
Nice job!

I'd just replace the following lines with (use `.pop()`, longer / more explicit variable names, `_` for unused element):
```Python
    for field in self.fields_desc:
        fname = field.name
        try:
            value = fields.pop(fname)
        except KeyError:
            continue
        self.fields[fname] = self.get_field(f).any2i(self, value)
    for fname, _ in fields:
        raise AttributeError(fname)
```